concent_api/core/message_handlers.py

```diff
@@ -1032,7 +1044,7 @@ def set_subtask_messages(
 
     for message_name, message_type in Subtask.MESSAGE_FOR_FIELD.items():
         message_to_store = subtask_messages_to_set.get(message_name)
-        if message_to_store is not None and getattr(subtask, message_name) is None:
+        if message_to_store is not None and (getattr(subtask, message_name) is None or replace == message_to_store):
```

Comment:
I see some downfalls of this mechanism, for example:
- you can replace only message at a time,
- you have to have original message which you are going to replace for comparison in this condition.

How about doing this:
- remove the `replace` parameter and related stuff.
- introduce a dict of lists in `Subtask` model which will define in which state which messages can be replaced (for validation). Lets call it `MESSAGE_REPLACEMENT_FOR_STATE`. For now it can only contain values for key `FORCING_REPORT`. Please remember about some extra asserts for this dict, see how they are done in other dicts in `Subtask` model.
- in the `if` above, instead of checking `replace == message_to_store` check if message type is in `MESSAGE_REPLACEMENT_FOR_STATE[subtask.state]`.

@cameel Should we keep in database old `report_computed_task` ? If not @kbeker please remember to remove it.

If not, what will happen if after assigning second `report_computed_task` you will call `subtask.report_computed_task` ? Which one will be called ?

Reply:
> Should we keep in database old `report_computed_task` ?

Won't hurt anything to keep it. We're keeping everything right now, right? It won't be used for processing but may help us see what actually happened when something goes wrong. So basically just for audit or debugging.

But it would be fine to remove it if keeping it is a problem.

> If not, what will happen if after assigning second `report_computed_task` you will call `subtask.report_computed_task` ? Which one will be called ?

The one the foreign key in the subtask points at. And that's what we want. We must never pull messages directly from `StoredMessage` because they don't have any inherent meaning. `StoredMessage` is meant to be just dumb storage. Only when a message is associated with something via a foreign key you can know it's the right message.

> introduce a dict of lists in Subtask model which will define in which state which messages can be replaced (for validation).

And here add an `assert` that does not let you change the foreign key if it's not `None` and is not listed in that dict.

Reply:
Done